## Content 

  - **Transpose of a dataframe**
  
  
  - **Joins in merge**
    - One-to-One
    - Many-to-One
    - Many-to-Many


  - **Conversion between numpy array and dataframe**


  - **Concatenation**
      - `df.append()`
  
  
  - **Use case: IMDB**
      - Handling Missing Values


In [ ]:
import numpy as np
import pandas as pd

***

## Transpose of dataframe

  - Change rows to cols and vice-versa
  - It can be done in a similar way as that of numpy

Syntax: `df.T` 

In [ ]:
df = pd.DataFrame({"A": [1, 10, 100],
                   "B": [2, 20, 200]
                   })
df

A    B
0    1    2
1   10   20
2  100  200

In [ ]:
df.T

0   1    2
A  1  10  100
B  2  20  200

***

## Joins in Merge

### 1. One-to-One join
  - Similar to concatenating columns

In [ ]:
df1 = pd.DataFrame({'employee': ['Bob', 'Jake', 'Lisa', 'Sue'],
                    'group': ['Accounting', 'Engineering', 'Engineering', 'HR']})
df2 = pd.DataFrame({'employee': ['Lisa', 'Bob', 'Jake', 'Sue'],
                    'hire_date': [2004, 2008, 2012, 2014]})
print(df1)
print(df2)

  employee        group
0      Bob   Accounting
1     Jake  Engineering
2     Lisa  Engineering
3      Sue           HR
  employee  hire_date
0     Lisa       2004
1      Bob       2008
2     Jake       2012
3      Sue       2014


In [ ]:
df3 = pd.merge(df1, df2)
df3

employee        group  hire_date
0      Bob   Accounting       2008
1     Jake  Engineering       2012
2     Lisa  Engineering       2004
3      Sue           HR       2014


  - The pd.merge() function recognizes that each DataFrame has an “employee” column, and automatically joins using this column as a key. 
  
  - The result of the merge is a new DataFrame that combines the information from the two inputs. 
  
  - The order of entries in each column is not necessarily maintained <br> In this case, the order of the “employee” column differs between df1 and df2, and the pd.merge() function correctly accounts for this.
  
  - Additionally, keep in mind that the merge in general discards the index, except in the special case of merges by index

### 2. Many to one join

  - One of the two key columns contains duplicate entries.
  - For the many-to-one case, the resulting DataFrame will preserve those duplicate entries as appropriate.


In [ ]:
df4 = pd.DataFrame({'group': ['Accounting', 'Engineering', 'HR'],
                           'supervisor': ['Carly', 'Guido', 'Steve']})

In [ ]:
df3

employee        group  hire_date
0      Bob   Accounting       2008
1     Jake  Engineering       2012
2     Lisa  Engineering       2004
3      Sue           HR       2014

In [ ]:
df4

group supervisor
0   Accounting      Carly
1  Engineering      Guido
2           HR      Steve

In [ ]:
pd.merge(df3, df4)

employee        group  hire_date supervisor
0      Bob   Accounting       2008      Carly
1     Jake  Engineering       2012      Guido
2     Lisa  Engineering       2004      Guido
3      Sue           HR       2014      Steve

### 3. Many to Many joins

  -  If the key column in both the left and right array contains duplicates, then the result is a many-to-many merge.
  - Lets look at an example to understand this better

In [ ]:
df5 = pd.DataFrame({'group': ['Accounting', 'Accounting',
                                     'Engineering', 'Engineering', 'HR', 'HR'],

                           'skills': ['math', 'spreadsheets', 'coding', 'linux',
                                      'spreadsheets', 'organization']})


In [ ]:
df5

group        skills
0   Accounting          math
1   Accounting  spreadsheets
2  Engineering        coding
3  Engineering         linux
4           HR  spreadsheets
5           HR  organization

In [ ]:
df1

employee        group
0      Bob   Accounting
1     Jake  Engineering
2     Lisa  Engineering
3      Sue           HR

In [ ]:
pd.merge(df1, df5)

employee        group        skills
0      Bob   Accounting          math
1      Bob   Accounting  spreadsheets
2     Jake  Engineering        coding
3     Jake  Engineering         linux
4     Lisa  Engineering        coding
5     Lisa  Engineering         linux
6      Sue           HR  spreadsheets
7      Sue           HR  organization

***


## Converting b/w Numpy Arrays and Pandas DataFrames


Let's **create a numpy matrix first**

#### Remember what` randint()` does?

- Generate random integers b/w provided range

In [ ]:
arr = np.random.randint(10, 100, size=(6,4))
arr

array([[43, 73, 50, 97],
       [49, 88, 88, 70],
       [99, 42, 85, 98],
       [16, 66, 53, 51],
       [70, 76, 58, 13],
       [95, 22, 17, 95]])

- We got a $6\times4$ **numpy matrix**


#### Now, To convert it to a pandas DataFrame

- All we have to do is **use `pd.DataFrame()`**

In [ ]:
df = pd.DataFrame(data=arr)
df

0   1   2   3
0  43  73  50  97
1  49  88  88  70
2  99  42  85  98
3  16  66  53  51
4  70  76  58  13
5  95  22  17  95

- It has **created a DataFrame with 6 rows and 4 columns**

#### Now, Let's see how we can convert a pandas DataFrame to numpy array

- We'll use the created DataFrame `df`


- We'll convert this DataFrame back to a numpy array


- All we have to do is **use `df.values`**

In [ ]:
df.values

array([[43, 73, 50, 97],
       [49, 88, 88, 70],
       [99, 42, 85, 98],
       [16, 66, 53, 51],
       [70, 76, 58, 13],
       [95, 22, 17, 95]])

#### So, very quickly we can convert b/w Numpy arrays and Pandas DataFrames

#### Be cautious though!

- This **works only on numerical values**


- As you know, **numpy CANNOT handle string values**

#### What if we try to convert a DataFrame having string values to Numpy array?


In [ ]:
df[0][0] = 'abc'
df

0   1   2   3
0  abc  73  50  97
1   49  88  88  70
2   99  42  85  98
3   16  66  53  51
4   70  76  58  13
5   95  22  17  95

In [ ]:
df.values

array([['abc', 73, 50, 97],
       [49, 88, 88, 70],
       [99, 42, 85, 98],
       [16, 66, 53, 51],
       [70, 76, 58, 13],
       [95, 22, 17, 95]], dtype=object)

#### Can you tell what happened here?

- When `df.values` encountered a string value


- It **converted all values to type `object`** (string)


#### So, be careful while doing conversions b/w Numpy Arrays and Pandas DataFrames


***

## Concatenation using df.append()

There also exists a shorter method of appending 1 dataframe to the other

This is through the `append()` method

Concatenation takes place only through axis = 0 

In [ ]:
df1

employee        group
0      Bob   Accounting
1     Jake  Engineering
2     Lisa  Engineering
3      Sue           HR

In [ ]:
df2

employee  hire_date
0     Lisa       2004
1      Bob       2008
2     Jake       2012
3      Sue       2014

In [ ]:
df1.append(df3, ignore_index = False)

employee        group  hire_date
0      Bob   Accounting        NaN
1     Jake  Engineering        NaN
2     Lisa  Engineering        NaN
3      Sue           HR        NaN
0      Bob   Accounting     2008.0
1     Jake  Engineering     2012.0
2     Lisa  Engineering     2004.0
3      Sue           HR     2014.0

In [ ]:
df1

employee        group
0      Bob   Accounting
1     Jake  Engineering
2     Lisa  Engineering
3      Sue           HR

#### How is it different from pd.concat() ?
  - The append() method does not modify the orginial object
  - It creates a new one with combined data
  - Only works along axis = 0
  - Can be used to concatenate only 2 dataframes/series at a time
  - Hence, it is not a very efficient method
 

***

## Dealing with Missing values in IMDB dataset




In [ ]:
!gdown 1UoooIoWfUFswDuPndOpHOLYE52dSRwCh

Downloading...
From: https://drive.google.com/uc?id=1UoooIoWfUFswDuPndOpHOLYE52dSRwCh
To: H:\Scaler work\dsml-course\07-08-09-Pandas\data.csv

  0%|          | 0.00/123k [00:00<?, ?B/s]
100%|##########| 123k/123k [00:00<00:00, 615kB/s]
100%|##########| 123k/123k [00:00<00:00, 612kB/s]


In [ ]:
data = pd.read_csv('data.csv',index_col=0)

In [ ]:
data


id  popularity                                     title  \
0     43597         150                                    Avatar   
1     43598         139  Pirates of the Caribbean: At World's End   
2     43599         107                                   Spectre   
3     43600         112                     The Dark Knight Rises   
4     43602         115                              Spider-Man 3   
...     ...         ...                                       ...   
1460  48363           3                            The Last Waltz   
1461  48370          19                                    Clerks   
1462  48375           7                                   Rampage   
1463  48376           3                                   Slacker   
1464  48395          14                               El Mariachi   

      vote_average  vote_count  year month       day      director_name  \
0              7.2       11800  2009   Dec  Thursday      James Cameron   
1              6.9        4500  2007   May  Saturday     Gore Verbinski   
2              6.3        4466  2015   Oct    Monday         Sam Mendes   
3              7.6        9106  2012   Jul    Monday  Christopher Nolan   
4              5.9        3576  2007   May   Tuesday          Sam Raimi   
...            ...         ...   ...   ...       ...                ...   
1460           7.9          64  1978   May    Monday    Martin Scorsese   
1461           7.4         755  1994   Sep   Tuesday        Kevin Smith   
1462           6.0         131  2009   Aug    Friday           Uwe Boll   
1463           6.4          77  1990   Jul    Friday  Richard Linklater   
1464           6.6         238  1992   Sep    Friday   Robert Rodriguez   

     gender  revenue_Mdollars  budget_Mdollars  
0      Male           2787.97           237.00  
1      Male            961.00           300.00  
2      Male            880.67           245.00  
3      Male           1084.94           250.00  
4      Male            890.87           258.00  
...     ...               ...              ...  
1460   Male              0.32             0.00  
1461   Male              3.15             0.03  
1462   Male              0.00             0.00  
1463   Male              0.00             0.00  
1464    NaN              2.04             0.22  

[1465 rows x 12 columns]



- We can check **`data.info()`** to get an **idea of distribution of missing values**


- It gives **no. of non-null (Available) values** in each column

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                1465 non-null   int64  
 1   popularity        1465 non-null   int64  
 2   title             1465 non-null   object 
 3   vote_average      1465 non-null   float64
 4   vote_count        1465 non-null   int64  
 5   year              1465 non-null   int64  
 6   month             1465 non-null   object 
 7   day               1465 non-null   object 
 8   director_name     1465 non-null   object 
 9   gender            1341 non-null   object 
 10  revenue_Mdollars  1465 non-null   float64
 11  budget_Mdollars   1465 non-null   float64
dtypes: float64(3), int64(4), object(5)
memory usage: 148.8+ KB


#### Notice the dataset

- Some of the values are missing - `NaN` in gender

#### But other than that is there any other missing values present ?

Lets check for this -

In [ ]:
data.loc[(data['revenue_Mdollars'] == 0.0) | (data['budget_Mdollars'] == 0.0)]

id  popularity              title  vote_average  vote_count  year  \
195   43918          16       The Campaign           5.6         578  2012   
271   44050          25   The Pink Panther           5.6         550  2006   
281   44072          20           The Edge           6.7         349  1997   
314   44133          13  Anna and the King           6.4         197  1999   
341   44181          13               Wolf           6.0         216  1994   
...     ...         ...                ...           ...         ...   ...   
1457  48323           5         The Mighty           7.1          51  1998   
1459  48359           2  George Washington           6.4          36  2000   
1460  48363           3     The Last Waltz           7.9          64  1978   
1462  48375           7            Rampage           6.0         131  2009   
1463  48376           3            Slacker           6.4          77  1990   

     month        day       director_name gender  revenue_Mdollars  \
195    Aug   Thursday           Jay Roach   Male            104.91   
271    Jan  Wednesday          Shawn Levy   Male              0.00   
281    Sep   Saturday        Lee Tamahori   Male             43.31   
314    Dec   Thursday        Andy Tennant   Male              0.00   
341    Jun     Friday        Mike Nichols   Male              0.00   
...    ...        ...                 ...    ...               ...   
1457   Oct     Friday       Peter Chelsom   Male              0.00   
1459   Oct     Sunday  David Gordon Green   Male              0.00   
1460   May     Monday     Martin Scorsese   Male              0.32   
1462   Aug     Friday            Uwe Boll   Male              0.00   
1463   Jul     Friday   Richard Linklater   Male              0.00   

      budget_Mdollars  
195               0.0  
271               0.0  
281               0.0  
314              75.0  
341               0.0  
...               ...  
1457              0.0  
1459              0.0  
1460              0.0  
1462              0.0  
1463              0.0  

[227 rows x 12 columns]

#### What you can refer from above ?

Usually,  Missing value in data can appear in 3 ways

    - NaN
    - NAN
    - nan

#### But that does not always hold true.

As you can see movies with `zero budget` or `zero revenue` is not possible, that are nothing but the missing values here 

So you have to be intuitive to decide for missing values in the dataset

Lets convert all missing values in columns to nan for better understanding

In [ ]:
data['revenue_Mdollars']=data['revenue_Mdollars'].replace(0, np.nan)
data['budget_Mdollars']=data['budget_Mdollars'].replace(0, np.nan)



#### Why do we need to handle these missing values?

- Because most ML and DS algorithms break when they encounter missing data


- Missing data depreciate the performance of our models


#### Question: Now, How can we deal with missing values? What ideas do you have?

- It is on **case-to-case basis**


- We have to **pick a method based on the dataset and SITUATION**


- We have to **check what will work and what not**


#### Ask yourself: What makes sense and what not?

- **DON'T just remove OR replace with default value OR replace with mean OR anything else blindly**


- The way we choose to deal with missing values can be **easily misleading**


#### So, we need to be very careful about how we choose to deal with missing data

- Use something **simple**


- But it **should make sense in the given situation**


### There are many ways to deal with missing values


In [ ]:
from numpy import NaN, NAN, nan

In [ ]:
# One interesting thing about Nan is this:

nan == nan

False

#### Why did it come out be `False`?

#### Can you compare two infinite or non-existent values?

- NO

- So, **be careful** while searching for missing values **using `==`**


#### So, How can we find if a value is missing or not?

- We can use **`is`**


- We can use **Pandas in-built function `isnull()`**

In [ ]:
nan is nan

True

In [ ]:
pd.isnull(nan)

True

#### Now coming back to our data
### How to find the total number of NULL values in each column:


In [ ]:
data.isnull().sum()

id                    0
popularity            0
title                 0
vote_average          0
vote_count            0
year                  0
month                 0
day                   0
director_name         0
gender              124
revenue_Mdollars    196
budget_Mdollars     137
dtype: int64

#### Now if you want to check for the percenatge of missing value for each column or features ?

In [ ]:

# len(movies) represents total number of rows
mis_val_per = data.isnull().sum()/len(data)*100
mis_val_per


id                   0.000000
popularity           0.000000
title                0.000000
vote_average         0.000000
vote_count           0.000000
year                 0.000000
month                0.000000
day                  0.000000
director_name        0.000000
gender               8.464164
revenue_Mdollars    13.378840
budget_Mdollars      9.351536
dtype: float64

#### What if you want to check for specific column gender ?

We can simply pass the column name as

In [ ]:
data['gender'].isnull().sum()

124

How to get the frequency of null values using `unique()`

- We can get **frequency of each value** in a column


- **No. of occurrences of each value** in a column


- Using `value_counts()`

In [ ]:
data['gender'].value_counts()

Male      1309
Female      32
Name: gender, dtype: int64

#### It tells that

- `Male` occurs 1309 times

- `Female` occurs 32 times




#### But its not telling the count of missing values

- Because by **default**, the **parameter `dropna` is set to `True`**


- `dropna=True` means it is **NOT going to count missing values**


- So, **we have to set `dropna=False`** to get the **count of missing values in a column**

In [ ]:
data['gender'].value_counts(dropna=False)

Male      1309
NaN        124
Female      32
Name: gender, dtype: int64

**Handling missing data**:

After finding the missing values, it needs to treat them.

There are many ways to handle missing values in the dataset

### Removing/Dropping the missing values

#### What if you have just 1 or very few rows which has missing data, compared to whole data?

- **Example**: only 10 rows out of 1 million rows having missing values


- We can simply remove those rows or columns using **`drop()`** 





### Replacing missing values with mean

But before making changes to the orginal data, lets again create a copy of it using copy()

In [ ]:
df=data.copy()

#### `any()`

#### How to filter rows that contains atleast one missing value

We can do it using `any()` method
 
 - The any() function is used to check whether any element is True, potentially over an axis.

In [ ]:
# Let's check the rows with null values before replace
df[df.isnull().any(axis=1)]

id  popularity                title  vote_average  vote_count  year  \
17    43620          42   The Golden Compass           5.8        1303  2007   
38    43653          65     Star Trek Beyond           6.6        2568  2016   
50    43669          90        Suicide Squad           5.9        7458  2016   
53    43672          44           Waterworld           5.9         992  1995   
81    43729          44  Wrath of the Titans           5.5        1431  2012   
...     ...         ...                  ...           ...         ...   ...   
1459  48359           2    George Washington           6.4          36  2000   
1460  48363           3       The Last Waltz           7.9          64  1978   
1462  48375           7              Rampage           6.0         131  2009   
1463  48376           3              Slacker           6.4          77  1990   
1464  48395          14          El Mariachi           6.6         238  1992   

     month       day       director_name gender  revenue_Mdollars  \
17     Dec   Tuesday         Chris Weitz    NaN            372.23   
38     Jul  Thursday          Justin Lin    NaN            343.47   
50     Aug   Tuesday          David Ayer    NaN            745.00   
53     Jul    Friday      Kevin Reynolds    NaN            264.22   
81     Mar   Tuesday  Jonathan Liebesman    NaN            301.00   
...    ...       ...                 ...    ...               ...   
1459   Oct    Sunday  David Gordon Green   Male               NaN   
1460   May    Monday     Martin Scorsese   Male              0.32   
1462   Aug    Friday            Uwe Boll   Male               NaN   
1463   Jul    Friday   Richard Linklater   Male               NaN   
1464   Sep    Friday    Robert Rodriguez    NaN              2.04   

      budget_Mdollars  
17             180.00  
38             185.00  
50             175.00  
53             175.00  
81             150.00  
...               ...  
1459              NaN  
1460              NaN  
1462              NaN  
1463              NaN  
1464             0.22  

[325 rows x 12 columns]

- `any()` method returns one value for each column, True if ANY value in that column is True, otherwise False.

- Here `.any()` checks if any value in a column is null


#### Now lets say you want to check for specific column say 'revenue'

In [ ]:
# Now check the null values only in revenue column
df.loc[df[["revenue_Mdollars"]].isna().any(axis=1)]


id  popularity                title  vote_average  vote_count  year  \
271   44050          25     The Pink Panther           5.6         550  2006   
314   44133          13    Anna and the King           6.4         197  1999   
341   44181          13                 Wolf           6.0         216  1994   
351   44198           9           Rollerball           3.4         106  2002   
374   44245          19      Mona Lisa Smile           6.5         393  2003   
...     ...         ...                  ...           ...         ...   ...   
1453  48294           6  She's Gotta Have It           6.1          25  1986   
1457  48323           5           The Mighty           7.1          51  1998   
1459  48359           2    George Washington           6.4          36  2000   
1462  48375           7              Rampage           6.0         131  2009   
1463  48376           3              Slacker           6.4          77  1990   

     month        day       director_name gender  revenue_Mdollars  \
271    Jan  Wednesday          Shawn Levy   Male               NaN   
314    Dec   Thursday        Andy Tennant   Male               NaN   
341    Jun     Friday        Mike Nichols   Male               NaN   
351    Feb     Friday      John McTiernan    NaN               NaN   
374    Dec     Friday         Mike Newell   Male               NaN   
...    ...        ...                 ...    ...               ...   
1453   Aug     Friday           Spike Lee   Male               NaN   
1457   Oct     Friday       Peter Chelsom   Male               NaN   
1459   Oct     Sunday  David Gordon Green   Male               NaN   
1462   Aug     Friday            Uwe Boll   Male               NaN   
1463   Jul     Friday   Richard Linklater   Male               NaN   

      budget_Mdollars  
271               NaN  
314              75.0  
341               NaN  
351               NaN  
374              65.0  
...               ...  
1453              NaN  
1457              NaN  
1459              NaN  
1462              NaN  
1463              NaN  

[196 rows x 12 columns]

#### Filling null values in revenue using mean:

lets first calculate the mean of column revenue

In [ ]:
m = df['revenue_Mdollars'].mean()
m

165.37986603624896

#### We can use `replace()` to replace the missing values with whatever we want, Let's say mean

Parameters:

- **`to_replace`** - What do I want to replace? 


- **`value`** - With what I want to replace?


- **`inplace=True`** - For permanent change

In [ ]:
df['revenue_Mdollars'].replace(to_replace=NaN, value=m, inplace=True)
df

id  popularity                                     title  \
0     43597         150                                    Avatar   
1     43598         139  Pirates of the Caribbean: At World's End   
2     43599         107                                   Spectre   
3     43600         112                     The Dark Knight Rises   
4     43602         115                              Spider-Man 3   
...     ...         ...                                       ...   
1460  48363           3                            The Last Waltz   
1461  48370          19                                    Clerks   
1462  48375           7                                   Rampage   
1463  48376           3                                   Slacker   
1464  48395          14                               El Mariachi   

      vote_average  vote_count  year month       day      director_name  \
0              7.2       11800  2009   Dec  Thursday      James Cameron   
1              6.9        4500  2007   May  Saturday     Gore Verbinski   
2              6.3        4466  2015   Oct    Monday         Sam Mendes   
3              7.6        9106  2012   Jul    Monday  Christopher Nolan   
4              5.9        3576  2007   May   Tuesday          Sam Raimi   
...            ...         ...   ...   ...       ...                ...   
1460           7.9          64  1978   May    Monday    Martin Scorsese   
1461           7.4         755  1994   Sep   Tuesday        Kevin Smith   
1462           6.0         131  2009   Aug    Friday           Uwe Boll   
1463           6.4          77  1990   Jul    Friday  Richard Linklater   
1464           6.6         238  1992   Sep    Friday   Robert Rodriguez   

     gender  revenue_Mdollars  budget_Mdollars  
0      Male       2787.970000           237.00  
1      Male        961.000000           300.00  
2      Male        880.670000           245.00  
3      Male       1084.940000           250.00  
4      Male        890.870000           258.00  
...     ...               ...              ...  
1460   Male          0.320000              NaN  
1461   Male          3.150000             0.03  
1462   Male        165.379866              NaN  
1463   Male        165.379866              NaN  
1464    NaN          2.040000             0.22  

[1465 rows x 12 columns]

* Now we can see in above results, the nan values at index 1462 and 1463 is **replaced with mean value** 165.37




- **`value=` can be anything**
    - mean
    - median
    - mode
    - 0
    - min, max
    - ...

Similarly, you can try to fill the missing values for column budget 



### Replace missing values with mode

After checking for numerical values, now lets see how you are going to handle missing values in categorical column `gender'

In [ ]:
df['gender'].value_counts(dropna=False)

Male      1309
NaN        124
Female      32
Name: gender, dtype: int64

In case of categorical we can replace the missing value by the most frequent value for that particular column, 

Here we replace the missing value by Male since the count of `Male` is more than `Female`

In [ ]:
#Take most occured category in that vairable (.mode())
Mode_Category = df['gender'].mode()
type(Mode_Category)

pandas.core.series.Series

Please note : `mode()` always returns Series even if only one value is returned.

So inorder to get first element of `mode()` we have to use index[0]

In [ ]:
Mode_Category = df['gender'].mode()[0]
Mode_Category

'Male'

Now replace NAN values with most occured category in actual variable

NOTE: Features having a max number of null values may create bias if replace with the most occurred category.
   

In [ ]:
 df['gender'].replace(to_replace=NaN, value=Mode_Category,inplace=True)
 df['gender'].value_counts(dropna=False)

Male      1433
Female      32
Name: gender, dtype: int64

In [ ]:
df

id  popularity                                     title  \
0     43597         150                                    Avatar   
1     43598         139  Pirates of the Caribbean: At World's End   
2     43599         107                                   Spectre   
3     43600         112                     The Dark Knight Rises   
4     43602         115                              Spider-Man 3   
...     ...         ...                                       ...   
1460  48363           3                            The Last Waltz   
1461  48370          19                                    Clerks   
1462  48375           7                                   Rampage   
1463  48376           3                                   Slacker   
1464  48395          14                               El Mariachi   

      vote_average  vote_count  year month       day      director_name  \
0              7.2       11800  2009   Dec  Thursday      James Cameron   
1              6.9        4500  2007   May  Saturday     Gore Verbinski   
2              6.3        4466  2015   Oct    Monday         Sam Mendes   
3              7.6        9106  2012   Jul    Monday  Christopher Nolan   
4              5.9        3576  2007   May   Tuesday          Sam Raimi   
...            ...         ...   ...   ...       ...                ...   
1460           7.9          64  1978   May    Monday    Martin Scorsese   
1461           7.4         755  1994   Sep   Tuesday        Kevin Smith   
1462           6.0         131  2009   Aug    Friday           Uwe Boll   
1463           6.4          77  1990   Jul    Friday  Richard Linklater   
1464           6.6         238  1992   Sep    Friday   Robert Rodriguez   

     gender  revenue_Mdollars  budget_Mdollars  
0      Male       2787.970000           237.00  
1      Male        961.000000           300.00  
2      Male        880.670000           245.00  
3      Male       1084.940000           250.00  
4      Male        890.870000           258.00  
...     ...               ...              ...  
1460   Male          0.320000              NaN  
1461   Male          3.150000             0.03  
1462   Male        165.379866              NaN  
1463   Male        165.379866              NaN  
1464   Male          2.040000             0.22  

[1465 rows x 12 columns]

***

Pandas is a very big library that is extensively used for data processing and analysis

While a lot of things have been covered in this lecture there are still other things that can be explored 